packages/ui/readme.md

```diff
@@ -1,28 +1,230 @@
 ```
 
-Finally, import the library CSS to the entry point of your app:
+Deprecated components can be imported from `@penumbra-zone/ui/components/ui/<component-name>`, where `<component-name>` should be replaced with the kebab-cased component name.
 
-```js
-import '@repo/ui/styles/globals.css';
+## Development
+
+These guidelines are for maintainers of the components in the Penumbra UI library.
+
+### Components must be located at `./components/v2/<ComponentName>/index.tsx`.
+
+This ensures that the Penumbra UI `package.json` `exports` field works as intended, so that components can be imported via `@penumbra-zone/ui/<ComponentName>`.
+
+Note that `<ComponentName>` should be replaced with the `UpperCamelCase` component name — e.g., `./components/v2/LoadingIndicator/index.tsx`.
+
+### Components that are only used within a parent component must sit in the parent component's directory, rather than be a sibling of the parent directory.
+
+```
+- src/components/
+  - HeaderWithDropdown/
+    - index.tsx
+    - Dropdown.tsx ✅ Correct, if Dropdown is only ever used inside HeaderWithDropdown
+```
+
+```
+- src/components/
+  - Dropdown.tsx ❌ Wrong, if Dropdown is only ever used inside HeaderWithDropdown
+  - HeaderWithDropdown/
+    - index.tsx
+```
+
+(One exception to this rule: if you're developing a component that will eventually be used by multiple other components, and just happens to be a child of only a single component at the moment, you can leave it as a sibling of its parent.)
+
+### Components should be located at the most specific possible directory level.
+
+For example, if the `Dropdown` component is used by both `HeaderWithDropdown` and `Menu` components, `Dropdown` should be placed in the lowest-level directory that contains both `HeaderWithDropdown` and `Menu`:
+
+```
+- src/components/
+  - SomeCommonParentOfBothHeaderWithDropdownAndMenu/
+    - index.tsx
+    - HeaderWithDropdown/
+      - index.tsx
+      - index.test.tsx
+    - Menu/
+      - index.tsx
+      - index.test.tsx
+    - Dropdown.tsx ✅ Correct - Dropdown is used by both Menu and HeaderWithDropdown, so it's a sibling to both
+```
+
+This, as opposed to e.g., placing it inside the `HeaderWithDropdown` directory (and then importing it from there in `Menu`), or inside a root-level directory. This way, components are nested as closely to the components using them as possible.
+
+```
+- src/components/
+  - SomeCommonParentOfBothHeaderWithDropdownAndMenu/
+    - index.tsx
+    - HeaderWithDropdown/
+      - index.tsx
+      - index.test.tsx
+      - Dropdown.tsx ❌ Wrong - Menu shouldn't be importing a child of HeaderWithDropdown
+    - Menu/
+      - index.tsx
+      - index.test.tsx
+```
+
+### Components must be the default export of the component file.
+
+This allows them to be imported as the default, rather than destructuring an object. It also encourages UI library maintainers to only export one component per file.
```

Comment:
Hm, a few thoughts:

- I don't have a strong opinion about this one way or the other, so I'm happy to change it.
- It seems that a lot of the discussions of this issue are claiming to be about named vs. default exports, when they _should_ be about named vs. unnamed exports. See Addy Osmani's tweet [here](https://x.com/addyosmani/status/1411233253948747777), which encourages people to name their exports (whether they are default exports or not). 
- I haven't had the experience that others are describing, where IDEs don't work well with default exports. In my experience with VSCode, if I want to import a default-exported `MyComponent`, I type `<MyComponent` in my JSX markup, and it automatically knows to import it. If I later rename `MyComponent`, it also updates everywhere in the codebase that imports `MyComponent`, even though it's the default export (since, after all, it is still a named function). It could just be that VSCode is especially smart about this, but I doubt it — I think it's a language-level thing, since `MyComponent` is actually a named function that happens to be exported as the default.

If you feel strongly about this, though, I can definitely remove this guideline.

Reply:
I've pushed https://github.com/penumbra-zone/web/pull/1411/commits/f01b8935ccb1fdfa695b4d38b999f7c4a259b8af to address this; but pending your response to my comment above, I can revert it back to my original guideline.